assets/js/util/tracking/index.js

```diff
@@ -0,0 +1,194 @@
+/**
+ * Event tracking utilities.
+ *
+ * Site Kit by Google, Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const DEFAULT_CONFIG = {
+	isFirstAdmin: false,
+	trackingEnabled: false,
+	trackingID: '',
+	referenceSiteURL: '',
+	userIDHash: '',
+};
+
+/**
+ * Tracking configuration.
+ */
+const config = { ...DEFAULT_CONFIG };
+
+/**
+ * The promise created when tracking is enabled.
+ * This reference is kept and returned for subsequent calls.
+ */
+let enableTrackingPromise;
+
+/**
+ * gtag script identifier.
+ */
+const SCRIPT_IDENTIFIER = 'data-googlesitekit-gtag';
+
+/**
+ * Data layer global used for internal/private Site Kit data.
+ */
+export const DATA_LAYER = '_googlesitekitDataLayer';
+
+/**
+ * Initializes tracking.
+ *
+ * @param {Object} _global The global object. Optional. (Used for testing only)
+ * @param {boolean} _resetPromise Clears the internal reference to the enableTrackingPromise (Optional - for testing only)
+ */
+export async function bootstrapTracking( _global = global, _resetPromise = false ) {
+	const {
+		isFirstAdmin,
+		trackingID,
+		trackingEnabled,
+		referenceSiteURL,
+		userIDHash,
+	} = _global._googlesitekitBase || DEFAULT_CONFIG;
+
+	Object.assign( config, DEFAULT_CONFIG, {
+		isFirstAdmin,
+		trackingID,
+		trackingEnabled,
+		userIDHash,
+		// Remove any trailing slash from the reference URL.
+		referenceSiteURL: referenceSiteURL.toString().replace( /\/+$/, '' ),
+	} );
+
+	if ( _resetPromise ) {
+		enableTrackingPromise = null;
+	}
+
+	return toggleTracking( trackingEnabled, _global );
+}
+
+/**
+ * Change the active state of tracking.
+ *
+ * @param {boolean} activeStatus The new state to set.
+ * @param {Object} _global The global object. Optional. (Used for testing only)
+ * @return {Promise} A promise that resolves with an object { trackingEnabled: (bool) }.
+ */
+export async function toggleTracking( activeStatus, _global = global ) {
```

Comment:
While the dependency injection of `global` seems neat, I don't like having that in a "public" function like here. _If_ we really need this, then let's assign and re-use the `_global` set in `bootstrapTracking()` and remove the parameter from the other functions.

However, I'm also curious: What does passing `_global` actually allows us to do that we wouldn't be able to do otherwise? Why not just access `global`?

Reply:
I don't think it offers additional capabilities, but is more of a pattern to make the usage more explicit by avoiding reaching outside of the function from within. That is, it's a bit closer to writing pure functions. I'm not sure if global contamination is a concern with Jest or not, but this avoids state leaking between tests by using a dedicated object instead of the real `global`.

To your point though, we can easily scope this to the module rather than each function so I'll update that. We're already doing that with the `enableTrackingPromise` and `config` so I don't see why we can't do that with `global` and avoid the extra parameter to each function.